### ISSUE TITLE
Most of the CLM library gets rebuilt for each test

### ISSUE THREAD

**Issue by billsacks** [ESMCI/cime#754]
@jedwards4b - This is basically just recording what we have already talked about. This is NOT critical, I just wanted to record this issue.

Starting with #697 (which by and large had some very good fixes), the CLM45/CLM50 library (which is supposed to be shared between tests) gets mostly rebuilt for each test. This can be seen in two ways:

1. Can check SHAREDLIB_BUILD times in test output: Should be much shorter for the second and subsequent tests of a given build configuration. Prior to #697 , these build times were indeed quite short for subsequent tests, but starting with #697 , the build times for subsequent tests are closer to those for the first test of a given build configuration.

    - Note that, for determining whether this issue has been resolved, it is good to check this both for mpi-serial and for non-mpi-serial cases, because at least in the past, there were some problems specific to mpi-serial in this respect.

2. Can check time stamps on files both in the shared build area (i.e., in sharedlibroot) and in the new 'install' area (i.e., in the case-specific bld directory) for the last case built in a large test suite: If everything is working right, the time stamps for the CLM build (the various *.o and *.mod files as well as the CLM library itself) should pre-date this case, because this case should just be reusing pre-existing builds.

Specifically, I'm checking these from a CLM tag, with:

```
export testid=`date +%m%d-%H%M`
export testroot=/glade/scratch/$USER/tests_${testid}
mkdir -p $testroot
ln -s $testroot
export category=aux_clm45
export compiler=intel
execca ./create_test --xml-category $category --xml-machine yellowstone --xml-compiler ${compiler} --test-id ${testid}.${category:7:2}.${compiler:0:1} --test-root ${testroot} --walltime 0:01 --no-run
```

From checking these things, I see that much of the CLM build is getting redone, starting with landunit_varcon. One guess (not investigated carefully) is that it's this line that's triggering the rebuild:

```
#include "shr_assert.h"
```

As far as I can tell, that header file itself is not updated (unless I'm missing something): it seems like it's getting the header file from `/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/cime/scripts/Tools/../../share/csm_share/include/shr_assert.h`. But is it possible that things are getting rebuilt simply because there are _directories_ in the include path with updated time stamps???

Some other things that I see have updated time stamps in the CLM build are Filepath, Deppath and Depends. Those seem to be the first files updated. Could that be the source of the problem?

In addition, all of the genf90 stuff seems to be getting redone unnecessarily - although that happens after landunit_varcon is rebuilt, so that seems to be independent of the landunit_varcon rebuild.

Here are some more details of what appears in the lnd bldlog:

The lnd bldlog starts with:

```
Resolved from env: CASEROOT
Resolved from env: CASEROOT
gmake complib -j 8 MODEL=clm COMPLIB=/glade/scratch/sacks/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/bld/intel/mpich2/debug/nothreads/MCT/noesmf/lib/libclm.a -f /glade/scratch/sacks/tests_1025-1124a/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/Tools/Makefile MACFILE=/glade/scratch/sacks/tests_1025-1124a/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/Macros.yellowstone :

 output:
 /glade/scratch/sacks/tests_1025-1124a/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/Tools/mkSrcfiles
cp -f /glade/scratch/sacks/tests_1025-1124a/sharedlibroot.1025-1124a.45.i/intel/mpich2/debug/nothreads/MCT/noesmf/clm/obj/Filepath /glade/scratch/sacks/tests_1025-1124a/sharedlibroot.1025-1124a.45.i/intel/mpich2/debug/nothreads/MCT/noesmf/clm/obj/Deppath
/glade/scratch/sacks/tests_1025-1124a/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/Tools/mkDepends  Deppath Srcfiles > /glade/scratch/sacks/tests_1025-1124a/sharedlibroot.1025-1124a.45.i/intel/mpich2/debug/nothreads/MCT/noesmf/clm/obj/Depends
/glade/scratch/sacks/tests_1025-1124a/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/Tools/mkSrcfiles
```

Then the first build is:

```
mpif90  -c -I. -I/glade/scratch/sacks/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/bld/intel/mpich2/debug/nothreads/include -I/glade/scratch/sacks/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/bld/intel/mpich2/debug/nothreads/MCT/noesmf/c1a1l1i1o1r1g1w1e1/include -I/glade/apps/opt/netcdf-mpi/4.3.3.1/intel/15.0.3/include -I/glade/apps/opt/pnetcdf/1.6.1/intel/15.0.3/include -I/glade/scratch/sacks/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/bld/intel/mpich2/debug/nothreads/include -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/cime/scripts/Tools/../../share/csm_share/shr -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/cime/scripts/Tools/../../share/csm_share/include -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/cime/scripts/Tools/../../share/shr_RandNum/include -I/glade/scratch/sacks/tests_1025-1124a/sharedlibroot.1025-1124a.45.i/intel/mpich2/debug/nothreads/MCT/noesmf/clm/obj -I. -I/glade/scratch/sacks/tests_1025-1124a/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/SourceMods/src.clm -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/main -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/biogeophys -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/biogeochem -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/soilbiogeochem -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/dyn_subgrid -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/init_interp -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/ED -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/ED/main -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/ED/biogeophys -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/ED/biogeochem -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/ED/fire -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/utils -I/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/cpl -I/glade/scratch/sacks/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.1025-1124a.45.i/bld/lib/include -no-opt-dynamic-align  -convert big_endian -assume byterecl -ftz -traceback -assume realloc_lhs -fp-model source     -xHost  -O0 -g -check uninit -check bounds -check pointers -fpe0 -check noarg_temp_created   -DLINUX -DCPRINTEL -DMCT_INTERFACE -DHAVE_MPI -DFORTRANUNDERSCORE -DNO_R16  -DINTEL_MKL -DHAVE_SSE2  -DHAVE_SLASHPROC -free  -DUSE_CONTIGUOUS=contiguous, /glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/main/landunit_varcon.F90
```

Then it does some genf90 stuff along with other builds; the first genf90 rerun is:

```
/glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/cime/scripts/Tools/../../externals/genf90/genf90.pl /glade/scratch/sacks/cesm_code/clm4_5_12_r198_cime_je/components/clm/src/utils/restUtilMod.F90.in > restUtilMod.F90
```

cc @ekluzek @bandre-ucar

**Comment by jedwards4b** [comment]
Here is what is going on, it's going to take a bit of work to fix it correctly:
The Filepath file is created for each case and includes the path to the SourceMods/src.clm directory for that case.   We didn't take this into account when we moved clm into sharedlib, since
each case has a different Filepath this update will cause a cascade of rebuilds to happen.   There is a potential problem with this - if a case, through user or test mods, puts a file in the SourceMods directory that file will be added to the library and every subsequent case matching that sharedlib path will get the source mod.   

My proposed fix would be to build the shared object files in the shared location as we are doing now (except without SourceMods/src.clm in the Filepath).  Then for each case use the case/bld/lnd/obj directory to write new Filepath - with sourcemods, then build a new library from the objects in sharedlib plus any new object found in Filepath.     I believe that this can be done with some modifications to the makefile, but I'm not going to take it on at this time.

**Comment by ekluzek** [comment]
Hi All

Good job on tracking this down!

I do think that SourceMods are incompatible with automated testing and
using a shared build. So here's a suggestion. You only build and use a
shared build if the SourceMods directories are empty. If SourceMods have
files in them, you do a standard non-shared build like other models. So the
clm build could flip between the shared build and a standard build for each
case based on the presence of SourceMod files.

Does that make it easier? Or harder?

*Erik Kluzek, (CGD at NCAR)*

*National Center for Atmospheric Research*

*Boulder CO, *

*(off) (303)497-1326 (fax) (303)497-1348*

*(skype) ekluzek    (cell) (303)859-0183*

*Pronouns: he/his/him*

*------------------ Home page ------------------------*

*          http://www.cgd.ucar.edu/~erik <http://www.cgd.ucar.edu/~erik>*

*          https://staff.ucar.edu/users/erik
<https://staff.ucar.edu/users/erik>*

*!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*

**Comment by billsacks** [comment]
@jedwards4b - I appreciate your taking the time to look into this.

However, I don't think this explains why this problem appeared after #697 . I just rechecked a test suite that was run using the older cime tag `clm4_5_12_r198_cime5.2.0-alpha.1-9-gd321e4b` (I did my testing using a CLM branch off of clm4_5_12_r199; the same results should apply for out-of-the-box clm4_5_12_r199). It doesn't look like any of the CLM library is getting rebuilt in that test suite. I checked various file modification times in `intel/mpich2/debug/nothreads/MCT/noesmf` and its subdirectories. The Filepath file _does_ get remade for each case in that older version of cime (as do the Deppath and Depends files), and the first line in Filepath points to SourceMods (e.g., `/glade/scratch/sacks/tests_1029-1605/SMS_D_Ld5.f45_f45.ICLM45ED.yellowstone_intel.clm-edTest.GC.1029-1605.45.i/SourceMods/src.clm`), but that doesn't seem to cause the rest of the CLM library to get rebuilt.

We recognized the SourceMods problem when the sharedlib build was first put in place, and partly because of that, we didn't allow SourceMods in testmods directories in cime4:

```perl
      if ($num_sourcemods_found > 0 && $is_test) {
         # Don't allow SourceMods in testmods, because (1) you shouldn't be
         # using SourceMods for automated tests, and (2) this likely won't
         # work correctly if multiple tests are sharing a single build.
         $logger->logdie ("SourceMods are not allowed in the testmods directory for automated tests");
      }
```

I'm not sure if that check has been migrated to cime5....

**Comment by billsacks** [comment]
The fix in #799 cuts the CLM yellowstone-intel test suite's build time from about 4 hours to about 2.5 hours. It still takes a little over a minute for each subsequent SHAREDLIB_BUILD, which presumably could be optimized in the future, but for now I'm happy. A couple of other things I checked (just for the record):

- No CLM files are rebuilt - great!

- It doesn't look like genf90 stuff is being redone unnecessarily for CLM any more - that problem is fixed

- Subsequent CLM builds redo the ar command to make libclm.a, but that just takes about 5 sec. per build
